### ISSUE TITLE
Pass JSDoc from parser rule to the generated interface

### ISSUE THREAD

**Issue by aabounegm** [eclipse-langium/langium#1673]
It would be great if the JSDoc comments above parser rules in grammar files are forwarded to the respective generated AST interface to improve IntelliSense when using that AST node type.
For example, if I write
```langium
/** A human being who has a name */
Person:
    'person' name=ID;
```
I would like `generated/ast.ts` to have:
```ts
/** A human being who has a name */
export interface Person extends AstNode {
    name: string;
}
```

It would also be nice if this can be extended to properties as well.

**Comment by aabounegm** [comment]
I tried to implement this myself using the `CommentProvider`, but I got stuck while trying to integrate it in `collectInferredTypes`
https://github.com/eclipse-langium/langium/blob/b2a6fe9684434117eb6a344fc2721e3c26d989bf/packages/langium/src/grammar/type-system/type-collector/inferred-types.ts#L259-L262

I am not sure how to associate the comment on the `rule` with the interface returned by `calculateInterfaces`

**Comment by dsogari** [comment]
It appears that comments are not present in the result from the [`parse`] method in `langium-parser.ts`. In that method, the `lexerResult` variable contains a list of `hidden` `IToken`s, each or which represents a comment in the grammar and has an `image` property containing the comment text.

I believe these items are being processed by the [`extractHiddenTokens`] method, but somehow they are not present in the produced [`result`]. I'm not sure how to proceed from this.

Related discussion: #1333 

[`parse`]: https://github.com/eclipse-langium/langium/blob/c3afb247d7dcfd41733c051b8db0395d50398b33/packages/langium/src/parser/langium-parser.ts#L234
[`extractHiddenTokens`]: https://github.com/eclipse-langium/langium/blob/c3afb247d7dcfd41733c051b8db0395d50398b33/packages/langium/src/parser/langium-parser.ts#L278
[`result`]: https://github.com/eclipse-langium/langium/blob/c3afb247d7dcfd41733c051b8db0395d50398b33/packages/langium/src/parser/langium-parser.ts#L242

**Comment by msujew** [comment]
@dsogari Hidden values are never visible in the AST, they are - as you correctly identified - extracted into the CST. You can use the `CommentProvider` or the `CstUtils.findCommentNode` function to get the associated comment node for an element.

**Comment by aabounegm** [comment]
Yes, that's what I'm trying to do, but I'm just not familiar enough with the relevant part of the code (yet). As far as I understand, the parsed Langium grammar passes through the [`collectAst`](https://eclipse-langium.github.io/langium/functions/langium_grammar.collectAst.html) function (which in turn calls other functions in the `type-collector` module) to transform it to an intermediate representation ([`AstTypes`](https://eclipse-langium.github.io/langium/types/langium_grammar.AstTypes.html)), which is then used to generate the TypeScript file (via `generate-ast` in [`langium-cli/src/generator/ast-generator.ts`](https://github.com/eclipse-langium/langium/blob/c3afb247d7dcfd41733c051b8db0395d50398b33/packages/langium-cli/src/generator/ast-generator.ts)).

@msujew please confirm if my understanding is correct.

Now, since the intermediate representation does not have any link to the original AST node, the JSDoc comment should be extracted during that transformation process and stored right on the relevant `AstType` object (`InterfaceType`, `UnionType`, and `Property`). Right now, I'm reading the code in `type-collector` to understand what should be added where. Though the code is readable, it's not heavily documented, so that might take some time 😅.
I'm thinking whether to add a `comment` field directly or perhaps `metadata` which could contain any additional metadata in the future (but I will probably keep it simple at least for the initial implementation).

**Comment by msujew** [comment]
Right, we lose a bunch of AST information during the transformation. It probably makes sense to keep the information somewhere and drag it through the whole type-collector. I wanted to take a look at this, but got sidetracked as well.

**Comment by dsogari** [comment]
I can see why this is not an easy task. There's indeed a transformation going on, but it's difficult to see a correspondence between AST nodes and the produced Interfaces just by looking at the code. Plus, I was quickly overwhelmed by the size of it. 😅 

On a side note, it would be nice to have the option to export the [Syntax Diagram] of the grammar. I was able to generate an HTML using the following piece of code (just replace `<lang>` with the language name):
<details>
<summary>Spoiler</summary>

```ts
import { EmptyFileSystem, createParser, LangiumParser } from 'langium';
import { createSyntaxDiagramsCode } from 'chevrotain';
import { create<lang>Services } from './src/language/<lang>-module.js';
import { writeFileSync } from 'fs';

class CustomLangiumParser extends LangiumParser {
  exportHtmlDiagram(): string {
    const rules = this.wrapper.getSerializedGastProductions();
    return createSyntaxDiagramsCode(rules);
  }
}

const { <lang> } = create<lang>Services(EmptyFileSystem);
const { definition } = <lang>.parser.Lexer;
const parser = new CustomLangiumParser(<lang>);
createParser(<lang>.Grammar, parser, definition).finalize();
const html = parser.exportHtmlDiagram();

if (process.argv.length > 2) {
  writeFileSync(process.argv[2], html);
} else {
  console.log(html);
}
```

</details>

[Syntax Diagram]: https://chevrotain.io/docs/guide/generating_syntax_diagrams.html
